File: 543-diameter-of-binary-tree/diameter-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var diameterOfBinaryTree = function(root) {

    let res = 0
    
    const dfs = (root) => {
        if (!root) {
            return -1
        }
        let left = dfs(root.left)
        let right = dfs(root.right)

        res = Math.max(res,left + right + 2)
        return 1 + Math.max(left, right)
    }
    dfs(root)
    return res
    
}

File: 543-diameter-of-binary-tree/diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.diameter = 0
        
        def dfs(node):
            if node is None:
                return
            left_tree = dfs(node.left) if node.left else 0
            right_tree = dfs(node.right) if node.right else 0
            self.diameter = max(self.diameter, left_tree + right_tree)
    
            return 1 + max(left_tree, right_tree)
        
        dfs(root)

        return self.diameter
        

        